### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] does not detect unnecessary non-null-assertion on a call expression

### ISSUE THREAD

**Issue by lvjiaxuan** [typescript-eslint/typescript-eslint#8141]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-type-assertion

### Description

I propose that the `no-unnecessary-type-assertion` rule should also check for the return type of a function.

### Fail

```TypeScript
declare function foo(): number;
const a = foo()!; // no error report
const b = foo() as number; // report error but bad fixed: `const b = foo() ; `
const c = <number>foo(); // good now
```

### Pass

```TypeScript
declare function foo(): number;
const a = foo();
const b = foo();
const c = foo();
```

### Additional Info

Im ready to send a PR after the issue is marked with the `accepting prs` label as indicated on the PR Checklist. Welcome to provide more test cases.

**Comment by auvred** [comment]
> const a = foo()!; // no error report

This is most likely a bug. When asserting a CallExpression we should check expresion's `callee`, not the `CallExpression` itself

**Comment by auvred** [comment]
> const a = foo()!; // no error report

This is most likely a bug, when asserting `CallExpression` we should check expression's `callee`, not `CallExpression` itself

https://github.com/typescript-eslint/typescript-eslint/blob/cde4d4d028ac9542a3372bf12c82ebfc137b5ae5/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts#L172

(because as far I know `CallExpression` can't have declarations)
https://github.com/typescript-eslint/typescript-eslint/blob/cde4d4d028ac9542a3372bf12c82ebfc137b5ae5/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts#L99-L103

---

> const b = foo() as number; // report error but bad fixed: `const b = foo() ; `

I think it's fine to file a separate issue about this, as it's not related to non-null assertions bug

**Comment by Josh-Cena** [comment]
> ```ts
> const b = foo() as number; // report error but bad fixed: `const b = foo() ; `
> ```

Why is this a bad fix? You need to run your formatter after autofixing; ESLint does not care about formatting and it's explicitly an antipattern for ESLint autofixers to conform to your formatting patterns.

**Comment by bradzacher** [comment]
> This is most likely a bug, when asserting CallExpression we should check expression's callee, not CallExpression itself

This is not correct. When you inspect the type of a call expression - you are inspecting the return type of the call. For example if you look at the types tab in the [playground](https://typescript-eslint.io/play#ts=5.3.2&showAST=types&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwAoBKALnlWQFsAjEGAbgCgw8BnDeKeAXgOIkAhI3gB6MZRzwGMHDHhwADvIysOXWnwGlu7SjXpNxk5apkw5C2sk1RgSLAA8QwCgAM2qTvC39EhLqi7ureXGDaADxUdAwAfAGCohLwAOaEDqg4AO7MzFBCzEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QeoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) and select the call expression - you will see it's type is `number`. So there is something else that's going on here to cause this bug.

**Comment by bradzacher** [comment]
@Josh-Cena i 100% agree with you! Though in this case it does appear to be an off-by-one error.

It's also done in a really dodgy way - by using "sketchy math" and assuming a specific code shape. We should really just be removing the `as` token to the end of the asserted type and then. Would be the same "bad fix" but would be less chance of a broken fix.

Tangent aside - I too am fine with us leaving this as is and assuming another rule will pickup the formatting, as is standard and expected in the ESLint ecosystem.

**Comment by Josh-Cena** [comment]
[Playground](https://typescript-eslint.io/play#ts=5.3.2&showAST=types&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwAoBKALnlWQFsAjEGAbgCgw8BnDeKeAXgOIkAhCyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by Josh-Cena** [comment]
Found https://github.com/typescript-eslint/typescript-eslint/pull/3235

**Comment by lvjiaxuan** [comment]
> > ```ts
> > const b = foo() as number; // report error but bad fixed: `const b = foo() ; `
> > ```
> 
> Why is this a bad fix? You need to run your formatter after autofixing; ESLint does not care about formatting and it's explicitly an antipattern for ESLint autofixers to conform to your formatting patterns.

Thanks for your opinion.
I just refer to the existing test case.

https://github.com/typescript-eslint/typescript-eslint/blob/cde4d4d028ac9542a3372bf12c82ebfc137b5ae5/packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts#L266-L282
